src/main/java/seedu/duke/command/SearchModuleCommand.java

```diff
@@ -0,0 +1,164 @@
+     * @param toSearchSemester the semester that user input
+     * @return true if module is offered in the semester
+     */
+    static boolean isOfferedInSemester(Module module, String toSearchSemester) {
+        // convert toSearchSemester to int
+        int toSearchSemesterInt = Integer.parseInt(toSearchSemester);
+
+        // check module is offered in which semester
+        if (module.getSemesterData(toSearchSemesterInt) != null) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Filter module by user input arguments and return a list of modules that match the search query.
+     * If no arguments are provided, no module will be returned.
+     * At least the module code or module title must be provided.
+     * If both module code and module title are provided, results will be display based on similar module code
+     * and module title but will not be repeated.
+     * Arguments can be in any order.
+     *
+     * @param toSearchModuleCode  the module code that user input
+     * @param toSearchLevel       the level that user input
+     * @param toSearchSemester    the semester that user input
+     * @param toSearchModuleTitle the module title that user input
+     * @return list of modules that match the search query
+     */
+    public static List<Module> filterModuleSearch(String toSearchModuleCode, String toSearchLevel,
+                                                  String toSearchSemester, String toSearchModuleTitle) {
+        List<Module> moduleList = Module.getAll();
+        List<Module> searchResult = new ArrayList<>();
+
+        // add all the mods with similar toSearchModuleCode to searchResult
+        for (Module m : moduleList) {
+            if (toSearchModuleCode != null && m.moduleCode.contains(toSearchModuleCode.toUpperCase())) {
+                searchResult.add(m);
+            }
+
+            if (toSearchModuleTitle != null && m.title.toLowerCase().contains(toSearchModuleTitle.toLowerCase())) {
+                // add only if it is not already in the list
+                if (!searchResult.contains(m)) {
+                    searchResult.add(m);
+                }
+            }
+        }
+
+        // filter the searchResult if toSearchLevel is not empty and level does not match
+        if (toSearchLevel != null) {
+            for (int i = 0; i < searchResult.size(); i++) {
+                if (!isSameModuleLevel(searchResult.get(i), toSearchLevel)) {
+                    searchResult.remove(i);
+                    i--;
+                }
+            }
+        }
+
+        // filter the searchResult if toSearchSemester is not empty and semester does not match
+        if (toSearchSemester != null) {
+            for (int i = 0; i < searchResult.size(); i++) {
+                if (!isOfferedInSemester(searchResult.get(i), toSearchSemester)) {
+                    searchResult.remove(i);
+                    i--;
```

Comment:
can be a bit dangerous to modify a list while iterating over it, usually good practice is to create a new list and add the elements matching the criteria to the new list.

Reply:
updated to good practice